pkg/upgrade/migrate.go

```diff
@@ -0,0 +1,359 @@
+	// by checking the existence of the path.
+	log.Info("checking host-local IPAM data dir dir existence...")
+	if _, err := os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("host-local IPAM data dir dir not found; no migration necessary, successfully exiting...")
+		return nil
+	}
+
+	// Get node resource to check for IPIP tunnel address.
+	log.Info("retrieving node for IPIP tunnel address")
+	node, err := c.Nodes().Get(ctxt, nodename, options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get calico node resource: %s", err)
+	}
+
+	// Migrate IPIP tunnel address if the field is empty.
+	if node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
+		log.Info("IPIP tunnel address not found, assigning...")
+
+		// Fetch k8s node.
+		k8sNode, err := k8sClient.CoreV1().Nodes().Get(nodename, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get k8s node resource: %s", err)
+		}
+
+		// Parse PodCIDR.
+		ip, cidr, err := net.ParseCIDR(k8sNode.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("PodCIDR %s did not parse successfully: %s", k8sNode.Spec.PodCIDR, err)
+		} else if cidr.Version() == 4 {
+			// We need to get the IP for the podCIDR and increment it to the
+			// first IP in the CIDR.
+			tunIp := ip.To4()
+			if tunIp == nil {
+				return fmt.Errorf("Cannot pick an IPv4 tunnel address from the given CIDR: %s", k8sNode.Spec.PodCIDR)
+			}
+			tunIp[3]++
+			node.Spec.BGP.IPv4IPIPTunnelAddr = tunIp.String()
+
+			// Assign the address via Calico IPAM.
+			ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+			if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+				IP:       *ipipTunnelAddr,
+				Hostname: nodename,
+				Attrs: map[string]string{
+					"node":           nodename,
+					"ipipTunnelAddr": "true",
+				},
+			}); err != nil {
+				if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+					return fmt.Errorf("failed to get add IPIP tunnel addr %s: %s", node.Spec.BGP.IPv4IPIPTunnelAddr, err)
+				}
+				log.Info("IPIP tunnel address already assigned, continuing...")
+			}
+
+			// Save the calico node object to the datastore.
+			if node, err = c.Nodes().Update(ctxt, node, options.SetOptions{}); err != nil {
+				return fmt.Errorf("failed to save newly updated node object: %s", err)
+			}
+
+			log.Info("Successfully assigned and updated IPIP tunnel address")
+		}
+	}
+	log.WithField("ip", node.Spec.BGP.IPv4IPIPTunnelAddr).Info("node has IPIP tunnel address")
+
+	// Open k8s-pod-directory to check for emptiness.
+	log.Info("checking if host-local IPAM data dir dir is empty...")
+	ipamDir, err := os.Open(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to open host-local IPAM data dir dir: %s", err)
+	}
+
+	// Check if the directory is empty.
+	if _, err = ipamDir.Readdirnames(1); err != nil {
+		if err == io.EOF {
+			log.Info("host-local IPAM data dir empty; no migration necessary, removing directory and exiting...")
```

Comment:
Doesn't look like we're actually deleting the dir here, unlike what the log suggests.